Conform .alterTable to the standard interface

xinbinhuang opened the pull request:
## Description

closes: #3281 

The `alterTable()` method on the SparkCatalog class always returns null. This violates the [API contract](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/connector/catalog/TableCatalog.html#alterTable-org.apache.spark.sql.connector.catalog.Identifier-org.apache.spark.sql.connector.catalog.TableChange...-) for the `TableCatalog` interface - this method must return the updated `Table` instance.

This PR fixes this by returning an updated `SparkTable` (which implements the `Table` interface) instance.


cc: @RussellSpitzer @netvl 

### Misc questions

I have two questions when implementing this
- Why do we have multiple versions of Spark in the repo. Do I need to copy the same code to different versions of Spark?
- What's the difference between the `SparkCatalog` and `SparkSessionCatalog`

xinbinhuang commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:
```diff
@@ -222,11 +222,11 @@ public SparkTable alterTable(Identifier ident, TableChange... changes) throws No
     try {
       Table table = load(ident);
       commitChanges(table, setLocation, setSnapshotId, pickSnapshotId, propertyChanges, schemaChanges);
+      table.refresh();
```
This doesn't seem to be necessary. The test still pass if I remove this. Should I remove it if it saves a network round trip to talk to the catalog?

rdblue replied:
This is unnecessary. The commit will update the table.

rdblue replied:
I should have been more clear. Can you please remove this?

rdblue commented:
> Why do we have multiple versions of Spark in the repo. Do I need to copy the same code to different versions of Spark?

You should make this change for 3.2. If anyone wants it for other versions, we can port the changes once they are reviewed.

> What's the difference between the SparkCatalog and SparkSessionCatalog

The `SparkSessionCatalog` is used to override Spark's singleton session catalog, `spark_catalog`.

rdblue commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkCatalogOperations.java:
```diff
@@ -0,0 +1,76 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
```
No. We don't automate licensing.

xinbinhuang replied:
Should we automate this using tools like pre-commit? I think it can help streamline the development process

RussellSpitzer replied:
IntelliJ will add this in automatically for you (or at least mine does)

rdblue replied:
No. We don't (and can't) automate licensing in anything that the project itself provides.

rdblue commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:
```diff
@@ -222,11 +222,11 @@ public SparkTable alterTable(Identifier ident, TableChange... changes) throws No
       Table table = load(ident);
       commitChanges(table, setLocation, setSnapshotId, pickSnapshotId, propertyChanges, schemaChanges);
+      table.refresh();
+      return new SparkTable(table, false);
```
Please add a comment to explain the boolean argument here.

xinbinhuang replied:
I set  `refreshEagerly` here to `false` because I think it makes sense to returns a table with the state right after the update. Otherwise, if I understand correctly, the table might be updated from other places unintentionally. And this will cause bugs on code that relies on the table state right after the update.

Besides, it seems that we always set this to false throughout the codebase atm.

- `cacheEabled` is `true` by default: https://github.com/apache/iceberg/blob/36e261bb40a0502185621165e0fa3441d4273d11/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java#L122
- https://github.com/apache/iceberg/blob/36e261bb40a0502185621165e0fa3441d4273d11/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java#L140
- https://github.com/apache/iceberg/blob/36e261bb40a0502185621165e0fa3441d4273d11/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java#L616

xinbinhuang replied:
I can still add an inline comment if you think that's needed

rdblue replied:
Yep, please do. We can follow up with a PR to fix the other case that you found passing `false`.

rdblue commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkCatalogOperations.java:
```diff
@@ -0,0 +1,75 @@
+    String fieldName = "location";
+    String propsKey = "note";
+    String propsValue = "jazz";
+    Table table = catalog.alterTable(
```
This PR returns the table instead of null, so the primary behavior that needs to be tested is that the table returned here is non-null. Can you add an assertion for that rather than dereferencing it later in other assertions?

rdblue commented:
Looks good overall, just a few minor things to fix. Thanks for picking this up, @xinbinhuang!

RussellSpitzer reviewed (approved):
No suggestions other than what @rdblue has already commented. Looks good to me

RussellSpitzer reviewed (approved):
This looks good to me, I think have SparkTable not refresh post commit change is fine although with some of our changes to the caching catalog the table would still automatically be updated within the next 30 seconds.

xinbinhuang commented:
@rdblue @RussellSpitzer , I've fixed the checkstyle error and change to `new SparkTable(table, true)`. After some thoughts, I changed it to `refreshEagerly=true` to align with the overall direction. PTAL.

rdblue commented on spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:
```diff
@@ -222,11 +222,10 @@ public SparkTable alterTable(Identifier ident, TableChange... changes) throws No
     try {
       Table table = load(ident);
       commitChanges(table, setLocation, setSnapshotId, pickSnapshotId, propertyChanges, schemaChanges);
+      return new SparkTable(table, true);
```
Can you add an inline comment to explain what the `true` argument is here?

kbendick commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkCatalogOperations.java:
```diff
@@ -0,0 +1,74 @@
+    Assert.assertEquals(table.schema().fields()[2], expectedField);
+
+    Assert.assertTrue(
+        "Created table missing property: " + propsKey,
```
Nit: This is an `alterTable` test, not create. Would you mind updating the assertion message?

Also, usually the assertion message should describe what went wrong and what should have happened.

So maybe something along the lines of `Adding a property to a table should return the updated table with the new property` or something.

kbendick commented on spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/TestSparkCatalogOperations.java:
```diff
@@ -0,0 +1,74 @@
+    Assert.assertTrue(
+        "Created table missing property: " + propsKey,
+        table.properties().containsKey(propsKey));
+    Assert.assertEquals("Property value is not the expected value",
```
Nit: For this one, the error message would possibly be more clear in describing what went wrong if there was a bit more explanation.

What do you think about `Altering a table to add a new property should add the correct value`. What do you think?

kbendick reviewed (approved):
Thank you for working on this @xinbinhuang! 

A few small nits, but nothing blocking and I’ll leave the changes at your discretion.

Would you mind updating the PR description about what the change actually changes / how it makes `alterTable` conform to the standard?

You can just copy it from the issue, but it helps for people who are going through commits and PRs in the future, instead of having to then go check the issue summary to see what the change is about.

Overall, this is great work!

xinbinhuang commented:
@kbendick Thanks for reviewing and the useful feedback! I've updated the PR with your suggestion. PTAL.

kbendick reviewed (approved):
LGTM

RussellSpitzer commented:
OK! Time to merge, thanks for the contribution @xinbinhuang !

xinbinhuang commented:
Thank you all for reviewing and giving feedback!
